include/seastar/core/abort_source.hh

```diff
@@ -104,10 +104,9 @@ private:
     std::exception_ptr _ex;
 
     void do_request_abort(std::optional<std::exception_ptr> ex) noexcept {
-        assert(_subscriptions);
-        _ex = ex.value_or(get_default_exception());
+        _ex = _ex ? _ex : ex.value_or(get_default_exception());
```

Comment:
How about `if (_ex) { return; }`?
This way walking the subscriptions would also happen exactly once.